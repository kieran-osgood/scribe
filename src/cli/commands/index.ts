// export * from './base-command';
// export * from './default-command';
export * from './init-command.js';
