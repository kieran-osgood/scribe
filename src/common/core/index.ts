// effect
export * as Boolean from '@effect/data/Boolean';
export * as Either from '@effect/data/Either';
export * from '@effect/data/Function';
export * as Option from '@effect/data/Option';
export * as Data from '@effect/data/Data';
// io
export * as Effect from '@effect/io/Effect';
export * as Runtime from '@effect/io/Runtime';
export * as Exit from '@effect/io/Exit';

//schema
export * as S from '@effect/schema/Schema';
export * as TF from '@effect/schema/TreeFormatter';

// export * as Sink from '@effect/stream/Sink';
// export * as Stream from '@effect/stream/Stream';

// core-ts
export * as IO from 'fp-ts/lib/IO.js';
export * as TE from 'fp-ts/lib/TaskEither.js';
export * as C from 'fp-ts/lib/Console.js';

// core-ts-std
export * as SStd from 'fp-ts-std/Struct';
