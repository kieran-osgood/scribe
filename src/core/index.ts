// effect
export * as Boolean from '@effect/data/Boolean';
export * as Chunk from '@effect/data/Chunk';
export * as Context from '@effect/data/Context';
export * as Data from '@effect/data/Data';
export { TaggedClass } from '@effect/data/Data';
export * as Either from '@effect/data/Either';
export * from '@effect/data/Function';
export * as Option from '@effect/data/Option';
export * as RA from '@effect/data/ReadonlyArray';
export * as R from '@effect/data/ReadonlyRecord';
export * as String from '@effect/data/String';
export * as T from '@effect/data/Tuple';
export * as Cause from '@effect/io/Cause';
export * as Layer from '@effect/io/Layer';
export * as Stream from '@effect/stream/Stream';

// io
export * as Effect from '@effect/io/Effect';
export * as Exit from '@effect/io/Exit';
export * as Runtime from '@effect/io/Runtime';

//schema
export * as S from '@effect/schema/Schema';
export * as TF from '@effect/schema/TreeFormatter';
