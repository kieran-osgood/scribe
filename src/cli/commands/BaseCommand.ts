import { Command, Option } from 'clipanion';
import * as t from 'typanion';
import { Effect, flow, pipe } from '@scribe/core';
import { runtimeDebug } from '@effect/data/Debug';
import * as FS from '@scribe/fs';
import * as Process from '../../process';

export abstract class BaseCommand extends Command {
  configPath = Option.String('-c,--config', 'scribe.config.ts', {
    description: 'Path to the config (default: scribe.config.ts)',
    validator: t.isString(),
  });

  test = Option.Boolean(
    '--test', //
    false,
    { hidden: true }
  );

  cwd = Option.String(
    '--cwd', //
    '',
    { hidden: true }
  );

  verbose = Option.Boolean('--verbose', false, {
    description: 'More verbose logging and error stack traces',
  });

  constructor() {
    super();
    runtimeDebug.minumumLogLevel = 'Info';
    runtimeDebug.tracingEnabled = false;

    if (process.env.NODE_ENV === 'development') {
      runtimeDebug.minumumLogLevel = 'Debug';
      runtimeDebug.tracingEnabled = true;
    }

    if (this.verbose) {
      runtimeDebug.minumumLogLevel = 'All';
    }
  }

  abstract executeSafe: () => Effect.Effect<
    Process.Process | FS.FS,
    unknown,
    void
  >;

  execute = (): Promise<void> => {
    return pipe(
      this.executeSafe(), //
      flow(
        this.test ? FS.FSMock : FS.FSLive,
        process.env.NODE_ENV === 'production' && !this.cwd
          ? Process.ProcessLive
          : this.cwd.length > 0
          ? Process.createMockProcess(this.cwd)
          : Process.MockProcess,

        Effect.runPromise,
        _ =>
          _.then(() => {
            this.context.stdout.write('Complete\n');
          }).catch(_ => {
            this.context.stdout.write(_.toString());
          })
      )
    );
  };
}
