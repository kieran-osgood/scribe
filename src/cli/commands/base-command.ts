import { runtimeDebug } from '@effect/data/Debug';
import { Cause, Context, Effect, pipe, Runtime } from '@scribe/core';
import { FS, Process } from '@scribe/services';
import { Command, Option } from 'clipanion';
import { Writable } from 'stream';
import * as t from 'typanion';

export abstract class BaseCommand extends Command {
  configPath = Option.String('-c,--config', 'scribe.config.ts', {
    description: 'Path to the config (default: scribe.config.ts)',
    validator: t.isString(),
  });
  test = Option.Boolean('--test', false, { hidden: true });
  cwd = Option.String('--cwd', '', { hidden: true });
  verbose = Option.Boolean('--verbose', false, {
    description: 'More verbose logging and error stack traces',
  });

  constructor() {
    super();

    if (process.env.NODE_ENV === 'production') {
      runtimeDebug.minumumLogLevel = 'Info';
      runtimeDebug.tracingEnabled = false;
    }

    if (this.verbose) {
      runtimeDebug.minumumLogLevel = 'All';
      runtimeDebug.tracingEnabled = true;
    }
  }

  abstract executeSafe: () => Effect.Effect<
    Process.Process | FS.FS,
    unknown,
    void
  >;

  // TODO: extract to module - replace manual service adding in tests
  private createContext = () =>
    pipe(
      Context.empty(),
      Context.add(Process.Process, Process.getProcess(this.cwd)),
      Context.add(FS.FS, FS.getFS(this.test)),
    );

  private handleExecutionResult =
    ({ stdout, verbose }: { stdout: Writable; verbose: boolean }) =>
    (commandEffect: Effect.Effect<Process.Process | FS.FS, unknown, void>) =>
      Effect.gen(function* ($) {
        const githubIssueUri =
          'https://github.com/kieran-osgood/scribe/issues/new';

        const result = yield* $(commandEffect, Effect.exit);

        if (result._tag === 'Failure') {
          const failOrCause = Cause.failureOrCause(result.cause);
          const errorWasManaged = failOrCause._tag === 'Left';

          if (!errorWasManaged || Runtime.isFiberFailure(failOrCause)) {
            stdout.write(
              `Unexpected Error
Please report this with the attached error: ${githubIssueUri}.`,
            );
          } else {
            stdout.write(
              `We caught an error during execution, this probably isn't a bug.
Check your 'scribe.config.ts', and ensure all files exist and paths are correct.

If you think this might be a bug, please report it here: ${githubIssueUri}.\n\n`,
            );
          }

          if (!verbose) {
            stdout.write(
              'You can enable verbose logging with --v, --verbose.\n\n',
            );
          }

          if (errorWasManaged) {
            if (
              !verbose &&
              Cause.isAnnotatedType(result.cause) &&
              Cause.isFailType(result.cause.cause)
            ) {
              let error =
                result.cause.cause.error?.toString() ?? 'Unable to parse error';

              if (isWithError(result.cause.cause.error)) {
                error = result.cause.cause.error.error.toString();

                // lol...look into any helpers to recursively pretty print
                if (isWithError(result.cause.cause.error.error)) {
                  error = result.cause.cause.error.error.error.toString();
                }
              }

              stdout.write(error);
            } else {
              stdout.write(Cause.pretty(result.cause));
            }
          }

          yield* $(
            Effect.sync(() => {
              if (process.env.NODE_ENV !== 'test') return process.exit(1);
            }),
          );
          return undefined;
        }

        return result.value;
      });

  execute = (): Promise<void> => {
    return pipe(
      this.executeSafe(),
      this.handleExecutionResult({
        stdout: this.context.stdout,
        verbose: this.verbose,
      }),
      Effect.provideContext(this.createContext()),
      Effect.runPromise,
    );
  };
}

type ContainsError = {
  error: Error;
};

const isWithError = (object: unknown): object is ContainsError => {
  return (
    Boolean(object) &&
    typeof object === 'object' &&
    Boolean((object as Record<string, unknown>)['error'])
  );
};
