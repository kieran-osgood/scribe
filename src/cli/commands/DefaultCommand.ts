import { Command, Option } from 'clipanion';
import * as t from 'typanion';

import { Effect, flow, pipe, RA } from '@scribe/core';
import { checkWorkingTreeClean } from '@scribe/git';

import { promptUserForMissingArgs } from '../../context';
import { BaseCommand } from './BaseCommand';
import { constructTemplate, Ctx, writeTemplate } from '../../templates';
import { Template } from '@scribe/config';
import { green } from 'colorette';

export class DefaultCommand extends BaseCommand {
  static override paths = [Command.Default];

  static override usage = Command.Usage({
    description: 'Scribe generates files based on mustache templates.',
    examples: [
      [`Interactively select template to use`, `$0`],
      [`Select via args`, `$0 --template screen --name Login`],
    ],
  });

  name = Option.String('-n,--name', {
    description: 'The key of templates to generate.',
    validator: t.isString(),
    required: false,
  });

  template = Option.String('-t,--template', {
    description:
      'Specify the name of the template to generate. Must be a key under templates in config.',
    validator: t.isString(),
    required: false,
  });

  overrideFlags(
    _: Effect.Effect.Success<ReturnType<typeof promptUserForMissingArgs>>
  ) {
    this.name = _.input.name;
    this.template = _.input.template;
    return _;
  }

  executeSafe = () =>
    pipe(
      // TODO: add ignore git?
      checkWorkingTreeClean(),

      Effect.flatMap(() =>
        promptUserForMissingArgs({
          name: this.name,
          template: this.template,
          configPath: this.configPath,
        })
      ),

      Effect.flatMap(_ => Effect.sync(() => this.overrideFlags(_))),

      Effect.flatMap(_ =>
        pipe(
          _.config.templates[_.input.template]?.outputs ?? [],
          RA.map(
            templateOutput => createTemplate({ templateOutput, ..._ }) //
          ),
          Effect.all,
          Effect.map(RA.flatten),
          Effect.flatMap(flow(Effect.all, Effect.flatMap(Effect.all)))
        )
      ),

      Effect.map(_ => {
        const results = pipe(
          _,
          RA.map(s => `- ${s}`),
          RA.join('\n')
        );
        this.context.stdout.write(green(`âœ…  Success!\n`));
        this.context.stdout.write(`Output files:\n${results}\n`);
      })
    );
}

/**
 * Wrap writeTemplate with Effect.gen
 * to keep track of the filePaths without process.cwd()
 * or return array of full paths and string.replace process.cwd()
 */
const createTemplate = (ctx: Ctx & { templateOutput: Template }) => {
  return pipe(
    Effect.gen(function* ($) {
      const templates = yield* $(constructTemplate(ctx));
      return pipe(templates, RA.map(Effect.map(writeTemplate)));
    })
  );
};
