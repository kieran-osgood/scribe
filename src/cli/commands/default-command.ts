import * as Config from '@scribe/config';
import { Template } from '@scribe/config';
import { Effect, flow, pipe, R, RA } from '@scribe/core';
import { Git, Process, Prompt } from '@scribe/services';
import { Command, Option } from 'clipanion';
import { green } from 'colorette';
import { PathOrFileDescriptor } from 'fs';
import path from 'path';
import * as t from 'typanion';

import { constructTemplate, Ctx, writeTemplate } from '../../templates';
import { BaseCommand } from './base-command';

export class DefaultCommand extends BaseCommand {
  static override paths = [Command.Default];

  static override usage = Command.Usage({
    description: 'Scribe generates files based on mustache templates.',
    examples: [
      [`Interactively select template to use`, `$0`],
      [`Select via args`, `$0 --template screen --name Login`],
    ],
  });

  name = Option.String('-n,--name', {
    description: 'The key of templates to generate.',
    validator: t.isString(),
    required: false,
  });

  template = Option.String('-t,--template', {
    description:
      'Specify the name of the template to generate. Must be a key under templates in config.',
    validator: t.isString(),
    required: false,
  });

  private rewriteFlagsWithUserInput(
    _: Effect.Effect.Success<
      ReturnType<
        InstanceType<typeof DefaultCommand>['promptUserForMissingArgs']
      >
    >,
  ) {
    this.name = _.input.name;
    this.template = _.input.template;
  }

  printOutput(_: PathOrFileDescriptor[]) {
    const results = pipe(
      _,
      RA.map(s => `- ${String(s)}`),
      RA.join('\n'),
    );
    this.context.stdout.write(green(`✅  Success!\n`));
    this.context.stdout.write(`Output files:\n${results}\n`);
  }

  promptUserForMissingArgs = () => {
    const _config = this.configPath;
    const _name = this.name;
    const _template = this.template;

    return Effect.gen(function* ($) {
      const configPath = yield* $(constructConfigPath(_config));
      const config = yield* $(Config.readConfig(configPath));
      const templates = yield* $(Config.readUserTemplateOptions(configPath));
      const input = yield* $(
        Prompt.launchPromptInterface({
          templates,
          flags: { name: _name, template: _template },
        }),
      );

      return { config, input, templateKeys: templates } as const;
    });
  };

  executeSafe = () =>
    pipe(
      // TODO: add ignore git
      Git.checkWorkingTreeClean(),
      Effect.flatMap(() => this.promptUserForMissingArgs()),
      Effect.tap(_ => Effect.sync(() => this.rewriteFlagsWithUserInput(_))),
      Effect.flatMap(createTemplates),
      Effect.map(_ => {
        const results = pipe(
          _,
          RA.map(s => `- ${String(s)}`),
          RA.join('\n'),
        );
        this.context.stdout.write(green(`✅  Success!\n`));
        this.context.stdout.write(`Output files:\n${results}\n`);
      }),
    );
}

/**
 * Constructs and writes templates to files persistently
 */
const createTemplate = (ctx: Ctx & { templateOutput: Template }) =>
  pipe(
    constructTemplate(ctx),
    Effect.map(RA.map(writeTemplate)),
    Effect.flatMap(Effect.all),
  );

const createTemplates = (ctx: Ctx) => {
  return pipe(
    R.get(ctx.input.template)(ctx.config.templates),
    Effect.map(_ => _.outputs),
    Effect.flatMap(
      flow(
        RA.map(
          templateOutput => createTemplate({ templateOutput, ...ctx }), //
        ),
        Effect.all,
        Effect.map(RA.flatten),
      ),
    ),
  );
};

const constructConfigPath = (filePath: string) =>
  pipe(
    Process.Process,
    Effect.flatMap(_process =>
      pipe(
        Effect.if(
          path.isAbsolute(filePath),
          Effect.succeed(filePath),
          Effect.succeed(path.join(_process.cwd(), filePath)),
        ),
      ),
    ),
  );
