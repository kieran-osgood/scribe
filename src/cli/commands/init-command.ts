import { Console, Inquirer } from '@scribe/adapters';
import { FS, Git, Process } from '@scribe/services';
import { Command } from 'clipanion';
import { Effect, pipe } from 'effect';
import path from 'path';

import { WARNINGS } from '../../common/constants';
import { BaseCommand } from './base-command';

export class InitCommand extends BaseCommand {
  static override paths = [['init']];

  static override usage = Command.Usage({
    description: 'Generates a scribe.config.ts file.',
  });

  // TODO: add flag to continue

  executeSafe = () => {
    return pipe(
      Console.logHeader('Init'),
      Effect.tap(() =>
        Console.logGroup('info', 'Git')('Checking working tree clean'),
      ),
      Effect.flatMap(() => Git.isWorkingTreeClean()),

      Effect.flatMap(
        Effect.if({
          onTrue: Effect.succeed(true),
          onFalse: Effect.gen(function* ($) {
            yield* $(Console.logWarn(WARNINGS.gitWorkingDirectoryDirty));
            return yield* $(Inquirer.continuePrompt());
          }),
        }),
      ),

      //  Print: Not in git repository?
      // prompt: Continue?
      // TODO: test this
      Effect.catchTag('SimpleGitError', () => Inquirer.continuePrompt()),

      Effect.flatMap(
        Effect.if({
          onTrue: pipe(
            Console.logGroup('info', 'Config')('Checking write path clear'),
            Effect.flatMap(() => checkConfigWritePathEmpty()),
            Effect.tap(() => Console.logInfo('Writing...')),
            Effect.flatMap(copyBaseScribeConfigToPath),
          ),
          onFalse: Effect.unit,
        }),
      ),

      Effect.catchTag('@scribe/core/fs/FileExistsError', error =>
        pipe(
          Console.logError(`Failed to create config. Path not empty.`),
          Effect.tap(() => Console.logFile(error.error.path.toString())),
        ),
      ),

      Effect.flatMap(fileDescriptor => {
        if (fileDescriptor) {
          return Console.logGroup(`success`, 'Success')().pipe(
            Effect.tap(() =>
              Console.logSuccess(
                'Scribe init complete. Edit the config to begin templating.',
              ),
            ),
            Effect.tap(() => Console.logFile(`${fileDescriptor.toString()}`)),
          );
        }

        return Effect.unit;
      }),
    );
  };
}

const createConfigPath = (_process: Process.Process) =>
  path.join(_process.cwd(), 'scribe.config.ts');

const checkConfigWritePathEmpty = () => {
  return Process.Process.pipe(
    Effect.flatMap(_process =>
      pipe(
        createConfigPath(_process),
        FS.isFileOrDirectory,
        Effect.if({
          onTrue: createFileExistsError(),
          onFalse: Effect.unit,
        }),
        Effect.catchTag('@scribe/core/fs/StatError', error => {
          /**
           * ENOENT indicates the path is clear, and we can safely write there
           */
          if (error.error.code === 'ENOENT') {
            return Effect.unit;
          }

          // TODO: add ignore file exists
          // TODO: test case that hits this?
          return Effect.fail(error);
        }),
      ),
    ),
  );
};

const copyBaseScribeConfigToPath = () =>
  pipe(
    FS.readFile('public/base.ts'),
    Effect.flatMap(configTxt =>
      Process.Process.pipe(
        Effect.map(createConfigPath),
        Effect.flatMap(_ => FS.writeFile(_, configTxt.toString(), null)),
      ),
    ),
  );

const createFileExistsError = () =>
  Process.Process.pipe(
    Effect.flatMap(_process =>
      Effect.fail(
        new FS.FileExistsError({
          error: new FS.AccessError({
            error: new Error(`${createConfigPath(_process)} already exists.`),
            path: createConfigPath(_process),
            mode: 0,
          }),
        }),
      ),
    ),
  );
