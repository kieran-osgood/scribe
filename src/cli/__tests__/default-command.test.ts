import { Effect, pipe } from '@scribe/core';
import { FS, Process } from '@scribe/services';
import * as memfs from 'memfs';
import { test } from 'vitest';

import { createConfigPathAbsolute } from '../commands';

const runEffect = ({
  effect,
  dir,
}: {
  effect: Effect.Effect<Process.Process | FS.FS, FS.StatError, string>;
  dir: string;
}) =>
  pipe(
    effect,
    Effect.provideService(Process.Process, Process.createProcessMock(dir)),
    Effect.provideService(FS.FS, FS.FSMock),
    Effect.runPromise,
  );

describe('createConfigPathAbsolute', () => {
  test('Should accept custom cwd', async () => {
    const dir =
      '/private/var/folders/vf/pdvmry710gncdjlmpr7pr97m0000gn/T/6babf399372540f63efac66bb6968305';
    memfs.vol.mkdirSync(dir, { recursive: true });

    const result = await runEffect({
      effect: createConfigPathAbsolute(dir),
      dir,
    });
    expect(result).toBe(`${dir}/scribe.config.ts`);
  });

  // test('Should 1', async () => {
  //   // const dir =
  //   //   '/private/var/folders/vf/pdvmry710gncdjlmpr7pr97m0000gn/T/6babf399372540f63efac66bb6968305';
  //   // memfs.vol.mkdirSync(dir, { recursive: true });
  //
  //   const dir = process.cwd();
  //   const result = await runEffect({
  //     effect: createConfigPathAbsolute('scribe.config.ts'),
  //     dir,
  //   });
  //   expect(result).toBe(`${dir}/scribe.config.ts`);
  // });
});
