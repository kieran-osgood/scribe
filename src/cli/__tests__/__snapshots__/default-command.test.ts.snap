// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DefaultCommand > [Given] Git Dirty > [Given] prompts to continue > [Then] user answers \`y\`, creates two files 2`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] Git Dirty > [Given] prompts to continue > [Then] user answers \`y\`, creates two files 3`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`DefaultCommand > [Given] Git clean > [Given] \`--config --template --name passed\` in > [Then] creates two files 2`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] Git clean > [Given] \`--config --template --name passed\` in > [Then] creates two files 3`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`DefaultCommand > [Given] Git clean > [When] \`--config\` passed in & fully interactive session > [Then] creates two files 2`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] Git clean > [When] \`--config\` passed in & fully interactive session > [Then] creates two files 3`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`DefaultCommand > [Given] Git clean > [When] \`--name\` passed in > [Then] creates two files 2`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] Git clean > [When] \`--name\` passed in > [Then] creates two files 3`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`DefaultCommand > [Given] Git clean > [When] \`--template\` passed in > [Then] creates two files 2`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] Git clean > [When] \`--template\` passed in > [Then] creates two files 3`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;
